staging/src/k8s.io/mount-utils/mount_linux.go

```diff
@@ -111,6 +112,47 @@ func (mounter *Mounter) hasSystemd() bool {
 	return *mounter.withSystemd
 }
 
+// Do a bind mount including the needed remount for applying the bind opts.
+// If the remount fails and we are running in a user namespace
+// figure out if the source filesystem has the ro, nodev, noexec, nosuid,
+// noatime, relatime or nodiratime flag set and try another remount with the found flags.
+func (mounter *Mounter) bindMountSensitive(mounterPath string, mountCmd string, source string, target string, fstype string, bindOpts []string, bindRemountOpts []string, bindRemountOptsSensitive []string, mountFlags []string, systemdMountRequired bool) error {
+	err := mounter.doMount(mounterPath, defaultMountCommand, source, target, fstype, bindOpts, bindRemountOptsSensitive, mountFlags, systemdMountRequired)
+	if err != nil {
+		return err
+	}
+	err = mounter.doMount(mounterPath, defaultMountCommand, source, target, fstype, bindRemountOpts, bindRemountOptsSensitive, mountFlags, systemdMountRequired)
+	if libcontaineruserns.RunningInUserNS() {
+		if err == nil {
+			return nil
+		}
+		// Check if the source has ro, nodev, noexec, nosuid, noatime, relatime,
+		// nodiratime flag...
+		var s unix.Statfs_t
+		if err := unix.Statfs(source, &s); err != nil {
+			return &os.PathError{Op: "statfs", Path: source, Err: err}
+		}
+		// ... and retry the mount with flags found above.
+		flagMapping := map[int]string{
+			unix.MS_RDONLY:     "ro",
```

Comment:
Could this cause unexpected behavior if "rw" was requested?

Reply:
It depends on your point of view. If you request a mount without the "rw" option being set it is mounted as read-write by default. Hence if you do not have the "rw" option among your mount options the code would have you end up with a read-only filesystem if the underlying filesystem is read-only. The same would happen if you set "rw" explicitly. You can argue that things should fail in this case instead of silently mounting it read-only.

Reply:
Thanks for writing it out. That was kinda the question. Should we just fail instead of "silently" falling back to ro

Reply:
If no option was specified by the requestor the purpose of the patch is to "fix" this silently. The patch could be changed to not apply options in the map where the requestor explicitly requested the opposite (e.g. do not apply ro if the requestor requested rw or do not apply nodev if dev was requested explicitly).